src/internal/components/dropdown/index.tsx

```diff
@@ -197,6 +197,14 @@ const Dropdown = ({
     } else {
       target.style.width = position.width;
     }
+
+    // Prevent the dropdown width from stretching beyond the trigger width
+    // if that is going to cause the dropdown to be cropped because of overflow
+    if (position.overflows) {
+      target.classList.remove(styles['stretch-beyond-trigger-width']);
```

Comment:
What is worrying is that the class is added and removed in different places. There is a chance that with another render of the dropdown the class is going to be re-added but not removed:

```
function DropdownConsumer () {
  const [counter, setCounter] = useState(0);
  useEffect(() => setInterval(() => setCounter(prev => prev + 1), 1000), []);
  return <Dropdown ... ><LargeContent /></Dropdown>
}
```

Also, what if the content changes? E.g. initially the dropdown renders a loading state and then the content appears. Does overflow calculation re-apply?

Reply:
You're right, there is an issue here because this `useLayoutEffect` hook has some dependencies so that it is called essentially when the dropdown changes from closed to open, but not if its content is changes while it's open. I think we will have to call `onDropdownOpen` (or at least part of it) essentially on every render. Here are two things to consider:

- Performance: the function call takes 2ms in average (on my machine), so I think this should not be a concern
- Behavior: should the dropdown placement change if its content changes? Since this hook only runs when opening, the current behavior is not to reposition the dropdown in any case while opened. I think this current behavior is good when select options are added or removed (which can happen with Autoselect) to prevent moving content that is already present in the page, but when changing the status from loading to content loaded, I think it would actually make sense to recalculate the position.

Reply:
I agree that the dropdown placement should stay the same but the concern is it might not to. The class that fixes the issue is only removed in the effect so the next render can bring it back and as result break the positioning. This can probably be fixed by applying the class from within the effect as well, wdyt?

Reply:
The code that (potentially) removes the class (and the maxWidth) is in a useLayoutEffect hook, so it will run after the component renders but before the browser paints the screen. I just moved this code to its own useLayoutEffect hook, because as said, we will have to check this on every render.

Reply:
Is there an advantage in using the layout effect with an empty dependencies array over computing the props in the render cycle, like:

```
const canOverflow = ...;
const maxWidth = canOverflow ? defaultMaxWidth : undefined;
const overflowClasses = canOverflow ? [styles['stretch-beyond-trigger-width']] : [];
```

Reply:
Ah, no - you do need to have the layout for that kind of computation of course.

Reply:
Exactly, we need to know what the rendered dimensions are going to be, so we need to render the component first...

Reply:
I am thinking on a slightly different solution then to not introduce the potential performance degradation. What if when the overflow is banned it is reflected in a new state so that the next renders will not override it?

```
useLayoutEffect(() => {
  if (!canOverflow) {
    setBanOveflow(true);
  }
}, [...]);

return <div style={{ maxWidth: !banOverflow ? defaultMaxWidth : undefined }} className={clsx(..., !banOverflow && styles['stretch-beyond-trigger-width'])} />
```

Alternatively, the state can be replaced with e.g. a new class name (like styles['ban-overflow']) to remove the effects in CSS. This class name can be added in the layout effect but then we need to ensure it persists over the next renders which can be achieved by also storing this info in a ref.

Reply:
Wouldn't the browser still do one paint with the layout that we don't want? Since this would "fix" the next render, but not the current one.

Reply:
Not sure about this one. When the state is changed from the layout effect it probably won't render the previous one. If yes, the alternative approach can be used or the styles and classes can be also removed imperatively.

Reply:
What you can also try is:

```
useLayoutEffect(() => {
  // ...
  if (!canOverflow) {
    banOverflowRef.current = true;
    dropdownRef.classList.add(styles['ban-overflow']);
  }
  // ...
}, [...]);

return <div ... className={clsx(..., banOverflowRef.current && styles['ban-overflow'])} />;
```

The above code ensures the `styles['ban-overflow']` is always present when we want to ban the overflow. In CSS you can use that class selector to negate the max-width and 'stretch-beyond-trigger-width' effects.